src/eachrow.jl

```diff
@@ -181,14 +182,186 @@ julia> df2 = @eachrow df begin
            :colX = cols(varB) == 2 ? pi * cols(varA) : cols(varB)
        end
 3×3 DataFrame
-│ Row │ A     │ B     │ colX    │
-│     │ Int64 │ Int64 │ Float64 │
-├─────┼───────┼───────┼─────────┤
-│ 1   │ 1     │ 2     │ 3.14159 │
-│ 2   │ 2     │ 1     │ 1.0     │
-│ 3   │ 3     │ 2     │ 9.42478 │
+ Row │ A      B      colX
+     │ Int64  Int64  Float64
+─────┼───────────────────────
+   1 │     1      2  3.14159
+   2 │     2      1  1.0
+   3 │     3      2  9.42478
+
+julia> x = [1, 1, 1];
+
+julia> @eachrow df begin
+           x[row] = :A
+       end;
+
+julia> x
+3-element Vector{Int64}:
+ 1
+ 2
+ 3
+
+julia> @eachrow df begin
+           @newcol m::Vector{Float64}
+           :m = mean(_DF[:, row])
+       end
+3×3 DataFrame
+ Row │ A      B      m
+     │ Int64  Int64  Float64
+─────┼───────────────────────
+   1 │     1      2  2.0
+   2 │     2      1  1.66667
+   3 │     3      2  1.22222
+
 ```
 """
 macro eachrow(d, body)
     esc(eachrow_helper(d, body, false))
 end
+
+function eachrow!_helper(df, body)
+    e_body, e_newcols = eachrow_find_newcols(body, Any[])
+    _df = gensym()
+    quote
+        let $_df = $df
+            local _N = nrow($_df)
+            local _DF = @transform!($_df, $(e_newcols...))
+            $(with_helper(:_DF, :(for row = 1:_N
+                $(eachrow_replace(e_body))
+            end)))
+            _DF
+        end
+    end
+end
+
+"""
+    @eachrow!(d, body)
+
+Act on each row of a data frame, similar to
+
+```
+for row in eachrow(d)
+    ... # Actions that modify `d`.
+end
+```
+
+Includes support for control flow and `begin end` blocks. Since the
+"environment" induced by `@eachrow df` is implicitly a single row of `d`,
+use regular operators and comparisons instead of their elementwise counterparts
+as in `@with`. Note that the scope within `@eachrow!` is a hard scope.
+
+`eachrow!` also supports special syntax for allocating new columns. The syntax
+`@newcol x::Vector{Int}` allocates a new uninitialized column `:x` with an `Vector` container
+with eltype `Int`.This feature makes it easier to use `eachrow` for data
+transformations. `_N` is introduced to represent the number of rows in the dataframe,
+`_DF` represents the `dataframe` including added columns, and `row` represents
+the index of the current row.
+
+Changes to the rows directly affect `d`. The operation will modify the
+data frame in place.
```

Comment:
I'd mention this earlier, otherwise the difference with `@eachrow` isn't obvious.

Reply:
added earlier, with a reference to `@eachrow`.